### ISSUE TITLE
What is the source for SCAN formula in code?

### ISSUE THREAD

**Issue by j-adamczyk** [networkit/networkit#1162]
SCAN Structural Similarity score is defined in the paper "SCAN: A Structural Clustering Algorithm for Networks" ([link](https://web.cs.ucla.edu/~yzsun/classes/2014Spring_CS7280/Papers/Clustering/SCAN.pdf)) as:
```
score(u, v) = |N(u) sum N(v)| / sqrt(deg(u) * deg(v))
```

In the paper "Demystifying Graph Sparsification Algorithms in Graph Properties Preservation" ([link](https://arxiv.org/pdf/2311.12314.pdf)), which relies on NetworKit for all implementations, it is instead defined as:
```
score(u, v) = (|N(u) intersection N(v)| + 1) / sqrt((deg(u) + 1) * (deg(v) + 1))
```

In NetworKit code ([code](https://github.com/networkit/networkit/blob/cc801fd8d593a8a70570a2d26f54b5ef53dba0b4/networkit/cpp/sparsification/SCANStructuralSimilarityScore.cpp#L7), it is defined as (if I can tell correctly):
```
score(u, v) = (num_triangles(e) + 1) / sqrt((deg(u) + 1) * (deg(v) + 1))
```

Why are there such differences in those formulas? In particular, why does NetworKit require triangles? This has a sizeable impact on computational requirements:
- SCAN original paper: O(|E|)
- SCAN sparsification paper: O(|E|)
- NetworKit, due to counting triangles: O(a(G) * |E|), where a(G) is the arboricity of the graph, as per [this link](https://networkit.github.io/dev-docs/cpp_api/classNetworKit_1_1TriangleEdgeScore.html) and paper "Triangle Listing Algorithms: Back from the Diversion" M. Ortmann, U. Brandes

**Comment by fabratu** [comment]
Hi,

thanks for the question. I was not part of the project, when this was implemented (nearly 10 years ago) - so this might not be the definitive answer. Looked at the paper resources, and your point is valid for the theoretical part. In practice, though, for many real world (complex) networks the arboricity is rather small, and the lower time complexity depends on the presence of certain data structures (e.g. hash tables) and assumptions. The graph data structure we use in NetworKit is likely not a good candidate for implementing an efficient overlap computation.

The following example compares the triangle counting algorithm from NetworKit, `std::set_intersection` from C++, a simple iterator for sorted lists and an `unordered_set`. All intersection algorithms are handled in parallel:

Generated BA graph (100k nodes):

```
[INFO ]: Needed 309ms for counting triangles
[INFO ]: Needed 142623ms for computing intersection with std::set_intersection.
[INFO ]: Needed 228883ms for computing intersection using iteration.
[INFO ]: Needed 690985ms for computing intersection using an unordered_set.
```

Although this is far from a thorough benchmark, triangle counting is - at least for these two instances and using the mentioned techniques - faster. The result may vary with other implementations / structured iterations, instances and number of CPU-cores, of course.

Used code:

```
#include <networkit/graph/Graph.hpp>
#include <networkit/graph/GraphTools.hpp>
#include <networkit/generators/BarabasiAlbertGenerator.hpp>
#include <networkit/io/NetworkitBinaryReader.hpp>
#include <networkit/linkprediction/NeighborhoodUtility.hpp>
#include <networkit/sparsification/SCANStructuralSimilarityScore.hpp>

#include <unordered_set>

struct Counter
{
  struct value_type { template<typename T> value_type(const T&) { } };
  void push_back(const value_type&) { ++count; }
  size_t count = 0;
};

int main() {

        auto G = NetworKit::BarabasiAlbertGenerator(50, 100000).generate();
        std::vector<NetworKit::count> overlap(G.numberOfEdges());

        G = NetworKit::GraphTools::toUndirected(G);
        G.removeSelfLoops();
        G.removeMultiEdges();
        G.indexEdges();
        G.sortEdges();

        auto triangleCounter = NetworKit::TriangleEdgeScore(G);
        triangleCounter.run();
        auto triangles = triangleCounter.scores();

        Aux::Timer intersectTimer;
        intersectTimer.start();

        G.parallelForEdges([&](NetworKit::node u, NetworKit::node v, NetworKit::edgeid eid) {

                // Similar to NetworKit::NeighborhoodUtility::getCommonNeighbors(G, u, v).size();
                // but this avoids multiple sorting calls for the same neighborhood.
                Counter c;
                std::vector<NetworKit::node> uNeighbors(G.neighborRange(u).begin(), G.neighborRange(u).end());
                std::vector<NetworKit::node> vNeighbors(G.neighborRange(v).begin(), G.neighborRange(v).end());
                std::set_intersection(uNeighbors.begin(), uNeighbors.end(), vNeighbors.begin(), vNeighbors.end(), std::back_inserter(c));
                overlap[eid] = c.count;
        });

        intersectTimer.stop();
        INFO("Needed ", intersectTimer.elapsedMilliseconds(), "ms for computing intersection with std::set_intersection.");

        Aux::Timer intersectTimer2;
        intersectTimer2.start();

        G.parallelForEdges([&](NetworKit::node u, NetworKit::node w, NetworKit::edgeid eid) {
                auto degW = G.degree(w);
                auto degU = G.degree(u);
                NetworKit::index currentU = 0;
                NetworKit::index currentW = 0;
                do {
                        auto idU = G.getIthNeighbor(u, currentU);
                        auto idW = G.getIthNeighbor(w, currentW);
                        if (idU == idW) {
                                overlap[eid]++;
                                currentU++;
                                currentW++;
                        } else if (idU < idW) {
                                currentU++;
                        } else {
                                currentW++;
                        }
                } while (currentU < degU && currentW < degW);
        });

          intersectTimer2.stop();
          INFO("Needed ", intersectTimer2.elapsedMilliseconds(), "ms for computing intersection using iteration.");

          Aux::Timer intersectTimer3;
          intersectTimer3.start();

          G.parallelForEdges([&](NetworKit::node u, NetworKit::node v, NetworKit::edgeid eid) {
                  std::unordered_set<NetworKit::node> neighborUnion;
                  neighborUnion.reserve(G.degree(u) + G.degree(v));

                  G.forNeighborsOf(u, [&](NetworKit::node w) {
                          neighborUnion.insert(w);
                  });

                  G.forNeighborsOf(v, [&](NetworKit::node w) {
                          bool ok = neighborUnion.insert(w).second;
                          if (!ok) overlap[eid]++;
                  });
          });

          intersectTimer3.stop();
          INFO("Needed ", intersectTimer3.elapsedMilliseconds(), "ms for computing intersection using an unordered_set.");

          return 0;
```

**Comment by j-adamczyk** [comment]
@fabratu it makes sense. But how about the actual values? I am not convinced that number of triangles is equal to either the neighbhorhood intersection or union. This is quite more problematic than computation time, since if this is true, then:
1) documentation of NetworKit is wrong, since it states the original SCAN paper (and should be more explicit about having slightly different approach)
2) paper relying on NetworKit has wrong formula and results, given the NetworKit implementation

I calculated two examples. The first one:
```
           ----- B ----
          /               \
E --- A                D --- F
          \               /
           ----- C ----
```

Computing SCAN for B-D edge:
- original SCAN formula:
```
3 / sqrt(2 * 3) = 3 / sqrt(6)
```
- SCAN formula from sparsification paper (assuming we don't include the node in its own neighborhood):
```
(0 + 1) / sqrt((2 + 1) * (3 + 1)) = 1 / sqrt(6)
```
- NetworKit formula:
```
(0 + 1) / 6 = 1 / sqrt(6)
```

If we look at another example, with a triangle this time:
```
A --- B --- D --- E
         \      /
           C
```

SCAN scores for edge B-D:
- original SCAN formula:
```
5 / sqrt(3 * 3) = 5 / 3
```
- SCAN formula from sparsification paper (again, we don't include the node in its own neighborhood):
```
(1 + 1) / sqrt((3+1) * (3 + 1)) = 2 / 4 = 1/2
```
- NetworKit formula:
```
(1 + 1) / sqrt((3+1) * (3 + 1)) = 2 / 4 = 1 / 2
```

So the results depend on the interpretation. NetworKit is not wrong per se, but this should be stated explicitly, and maybe also provide the citation of the sparsification paper. I'm also not exactly sure if triangle counting is equal to neighborhood intersection, but it's just me.

**Comment by fabratu** [comment]
Hm, sorry - so far I only looked at the papers, not your concrete formulas. 

Where did you get the `sum` (union?) from? In the original paper, the structural similarity is likely definition #2, which is the intersection of neighborhoods. In their definition of the neighborhood, they include the node itself. This is the same as what is done in the sparsification paper, where the neighborhood is just every node in distance 1. This then explains the `+1`. 

As for the general question, whether this is correct - simple [triangle counting algorithms](https://perso.liris.cnrs.fr/hamida.seba/DMKD-seba-2017.pdf) use neighborhood intersection to list all triangles.

**Comment by j-adamczyk** [comment]
So it turns out I made a mistake, and you're totally right, there is always an intersection. And combined with your explanation, now I get everything. Thanks a lot!

**Comment by fabratu** [comment]
Perfect, and I get the confusion. Still, the documentation is surely something, where we can improve. Will close the issue for now. Feel free to reopen it if there are more questions.
